// @stoplight/spectral-rulesets/dist/asyncapi/functions/asyncApi2OperationIdUniqueness.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const spectral_core_1 = require("@stoplight/spectral-core");
const getAllOperations_1 = require("./utils/getAllOperations");
exports.default = (0, spectral_core_1.createRulesetFunction)({
    input: {
        type: 'object',
        properties: {
            channels: {
                type: 'object',
                properties: {
                    subscribe: {
                        type: 'object',
                    },
                    publish: {
                        type: 'object',
                    },
                },
            },
        },
    },
    options: null,
}, function asyncApi2OperationIdUniqueness(targetVal, _) {
    const results = [];
    const operations = (0, getAllOperations_1.getAllOperations)(targetVal);
    const seenIds = [];
    for (const { path, operation } of operations) {
        if (!('operationId' in operation)) {
            continue;
        }
        const operationId = operation.operationId;
        if (seenIds.includes(operationId)) {
            results.push({
                message: '"operationId" must be unique across all the operations.',
                path: [...path, 'operationId'],
            });
        }
        else {
            seenIds.push(operationId);
        }
    }
    return results;
});
