// @stoplight/spectral-rulesets/dist/oas/functions/oasExample.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const isObject_1 = require("./utils/isObject");
const oasSchema_1 = (0, tslib_1.__importDefault)(require("./oasSchema"));
const spectral_core_1 = require("@stoplight/spectral-core");
const spectral_formats_1 = require("@stoplight/spectral-formats");
const MEDIA_VALIDATION_ITEMS = {
    2: [
        {
            field: 'examples',
            multiple: true,
            keyed: false,
        },
    ],
    3: [
        {
            field: 'example',
            multiple: false,
            keyed: false,
        },
        {
            field: 'examples',
            multiple: true,
            keyed: true,
        },
    ],
};
const SCHEMA_VALIDATION_ITEMS = {
    2: ['example', 'x-example', 'default'],
    3: ['example', 'default'],
};
function* getMediaValidationItems(items, targetVal, givenPath, oasVersion) {
    for (const { field, keyed, multiple } of items) {
        if (!(field in targetVal)) {
            continue;
        }
        const value = targetVal[field];
        if (multiple) {
            if (!(0, isObject_1.isObject)(value))
                continue;
            for (const exampleKey of Object.keys(value)) {
                const exampleValue = value[exampleKey];
                if (oasVersion === 3 && keyed && (!(0, isObject_1.isObject)(exampleValue) || 'externalValue' in exampleValue)) {
                    continue;
                }
                const targetPath = [...givenPath, field, exampleKey];
                if (keyed) {
                    targetPath.push('value');
                }
                yield {
                    value: keyed && (0, isObject_1.isObject)(exampleValue) ? exampleValue.value : exampleValue,
                    path: targetPath,
                };
            }
            return;
        }
        else {
            return yield {
                value,
                path: [...givenPath, field],
            };
        }
    }
}
function* getSchemaValidationItems(fields, targetVal, givenPath) {
    for (const field of fields) {
        if (!(field in targetVal)) {
            continue;
        }
        yield {
            value: targetVal[field],
            path: [...givenPath, field],
        };
    }
}
exports.default = (0, spectral_core_1.createRulesetFunction)({
    input: {
        type: 'object',
    },
    options: {
        type: 'object',
        properties: {
            oasVersion: {
                enum: [2, 3],
            },
            schemaField: {
                type: 'string',
            },
            type: {
                enum: ['media', 'schema'],
            },
        },
        additionalProperties: false,
    },
}, function oasExample(targetVal, opts, context) {
    const formats = context.document.formats;
    const schemaOpts = {
        schema: opts.schemaField === '$' ? targetVal : targetVal[opts.schemaField],
    };
    let results = void 0;
    const validationItems = opts.type === 'schema'
        ? getSchemaValidationItems(SCHEMA_VALIDATION_ITEMS[opts.oasVersion], targetVal, context.path)
        : getMediaValidationItems(MEDIA_VALIDATION_ITEMS[opts.oasVersion], targetVal, context.path, opts.oasVersion);
    if ((formats === null || formats === void 0 ? void 0 : formats.has(spectral_formats_1.oas2)) && 'required' in schemaOpts.schema && typeof schemaOpts.schema.required === 'boolean') {
        schemaOpts.schema = { ...schemaOpts.schema };
        delete schemaOpts.schema.required;
    }
    for (const validationItem of validationItems) {
        const result = (0, oasSchema_1.default)(validationItem.value, schemaOpts, {
            ...context,
            path: validationItem.path,
        });
        if (Array.isArray(result)) {
            if (results === void 0)
                results = [];
            results.push(...result);
        }
    }
    return results;
});
