// mutation-testing-elements/dist/components/state-filter/state-filter.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, html, unsafeCSS } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { repeat } from 'lit/directives/repeat.js';
import { bootstrap } from '../../style';
import style from './state-filter.scss';
import { createCustomEvent } from '../../lib/custom-events';
let FileStateFilterComponent = class FileStateFilterComponent extends LitElement {
    constructor() {
        super(...arguments);
        this.next = (ev) => {
            ev.stopPropagation();
            this.dispatchEvent(createCustomEvent('next', undefined, { bubbles: true, composed: true }));
        };
        this.previous = (ev) => {
            ev.stopPropagation();
            this.dispatchEvent(createCustomEvent('previous', undefined, { bubbles: true, composed: true }));
        };
    }
    updated(changedProperties) {
        if (changedProperties.has('filters')) {
            this.dispatchFiltersChangedEvent();
        }
    }
    checkboxChanged(filter, enabled) {
        filter.enabled = enabled;
        this.dispatchFiltersChangedEvent();
    }
    dispatchFiltersChangedEvent() {
        this.dispatchEvent(createCustomEvent('filters-changed', this.filters.filter(({ enabled }) => enabled).map(({ status }) => status)));
    }
    render() {
        return html `
      <div class="legend col-md-12 d-flex align-items-center">
        <div class="d-flex me-2">
          <button title="Previous" @click=${this.previous} class="me-1 btn btn-sm btn-secondary mte-previous" type="button">&lt;</button>
          <button title="Next" @click=${this.next} class="btn btn-sm btn-secondary mte-next" type="button">&gt;</button>
        </div>

        ${this.filters &&
            repeat(this.filters, 
            // Key function. I super duper want that all properties are weighed here,
            // see https://lit-html.polymer-project.org/guide/writing-templates#repeating-templates-with-the-repeat-directive
            (filter) => JSON.stringify(filter), (filter) => html `<div data-status="${filter.status}" class="form-check form-check-inline">
            <label class="form-check-label">
              <input
                class="form-check-input"
                type="checkbox"
                ?checked="${filter.enabled}"
                value="${filter.status}"
                @input="${(el) => this.checkboxChanged(filter, el.target.checked)}"
              />
              <span class="badge bg-${filter.context}">${filter.label} (${filter.count})</span>
            </label>
          </div>`)}
      </div>
    `;
    }
};
FileStateFilterComponent.styles = [bootstrap, unsafeCSS(style)];
__decorate([
    property({ type: Array })
], FileStateFilterComponent.prototype, "filters", void 0);
FileStateFilterComponent = __decorate([
    customElement('mte-state-filter')
], FileStateFilterComponent);
export { FileStateFilterComponent };
